File: dynamic-programming/rod-cutting/javascript/RodCutting.js
const price = [1, 5, 8, 9];
var size = price.length;

console.log("Max obtainable revenue: " + rodCut(price, size));

var arr = new Array(size);
console.log("Max obtainable revenue (top down): " + rodCutTopDown(price, size, arr));

var res = rodCutBottomUp(price);
console.log("Max obtainable revenue (bottom up): " + res);

//DP : Bottom Up Approach O(n^2)
function rodCutBottomUp(price) {
    let r = new Array(price.length + 1);
    r[0] = 0;
    let maxRevenue = Number.MIN_SAFE_INTEGER;

    for (let i = 1; i < r.length; i++) {
        for (let j = 1; j <= i; j++) {
            maxRevenue = Math.max(maxRevenue, price[j - 1] + r[i - j]);
        }
        r[i] = maxRevenue;
    }
    return r[price.length];
}

//DP : Top Down Approach O(n^2)
function rodCutTopDown(price, n, arr) {
    if (n == 0)
        return 0;

    if (arr[n - 1] > 0)
        return dp[n - 1];

    let maxRevenue = Number.MIN_SAFE_INTEGER;

    for (let i = 1; i <= n; i++)
        maxRevenue = Math.max(maxRevenue, price[i - 1] + rodCut(price, n - i));

    return arr[n - 1] = maxRevenue;
}

//Recursive Approach (exponential)
function rodCut(price, n) {

    if (n == 0)
        return 0;

    let maxRevenue = Number.MIN_SAFE_INTEGER;

    for (let i = 1; i <= n; i++)
        maxRevenue = Math.max(maxRevenue, price[i - 1] + rodCut(price, n - i));

    return maxRevenue;
}

File: dynamic-programming/rod-cutting/python/RodCutting.py
import sys


# Recursive Approach (exponential)
def rod_cut(price, n):
    if n == 0:
        return 0
    max_revenue = -sys.maxsize
    for i in range(1, n + 1):
        max_revenue = max(max_revenue, price[i - 1] + rod_cut(price, n - i))
    return max_revenue


# DP : Top Down Approach O(n^2)
def rod_cut_topdown(price, n, dp):
    if n == 0:
        return 0

    if dp[n - 1] > 0:
        return dp[n - 1]

    max_revenue = -sys.maxsize
    for i in range(1, n + 1):
        max_revenue = max(max_revenue, price[i - 1] + rod_cut_topdown(price, n - i, dp))
    dp[n - 1] = max_revenue
    return dp[n - 1]


# DP : Bottom Up Approach O(n^2)
def rod_cut_bottomup(price, n):
    revenues = [0] * (n + 1)
    # revenues[0] = 0

    max_revenue = -sys.maxsize
    for i in range(1, n + 1):
        for j in range(1, i + 1):
            max_revenue = max(max_revenue, price[j - 1] + revenues[i - j])
        revenues[i] = max_revenue
    return revenues[n]


if __name__ == '__main__':
    price = [4, 5, 8, 9]
    size = len(price)
    print('max revenue', rod_cut(price, size))

    dp = [0] * size
    print('max revenue (top down):', rod_cut_topdown(price, size, dp))

    print('max revenue (bottom up):', rod_cut_bottomup(price, size))
